packages/sources/tradingeconomics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,104 @@
+import { makeLogger, SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
+import includes from '../config/includes.json'
+
+import { customSettings } from '../config'
+import { PriceEndpointParams } from '@chainlink/external-adapter-framework/adapter'
+
+const logger = makeLogger('Tradingeconomics WS')
+
+export type EndpointTypes = {
+  Request: {
+    Params: PriceEndpointParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: Message[]
+  }
+}
+
+type Includes = {
+  from: string
+  to: string
+  includes: IncludePair[]
+}
+
+export type IncludePair = {
+  from: string // From symbol
+  to: string // To symbol
+  adapters?: string[] // Array of adapters this applies to
+  inverse?: boolean // If the inverse should be calculated instead
+  tokens?: boolean // If the token addresses should be used instead
+}
+
+interface Message {
+  s: string
+  i: string
+  pch: number
+  nch: number
+  bid: number
+  ask: number
+  price: number
+  dt: number
+  state: string
+  type: string
+  dhigh: number
+  dlow: number
+  o: number
+  prev: number
+  topic: string
+}
+
+const withApiKey = (url: string, key: string, secret: string) => `${url}?client=${key}:${secret}`
+const getSubscription = (to: string) => ({ topic: 'subscribe', to })
+
+const baseFromIncludes = includes.reduce(
+  (basesMap: { [from: string]: string }, includesSet: Includes) => {
+    const { includes } = includesSet
+    for (const includePair of includes) {
+      basesMap[includePair.from] = includesSet.from
+    }
+    return basesMap
+  },
+  {},
+)
+
+export const wsTransport: WebsocketReverseMappingTransport<EndpointTypes, string> =
+  new WebsocketReverseMappingTransport<EndpointTypes, string>({
+    url: (context) => {
+      const { API_CLIENT_KEY, API_CLIENT_SECRET, WS_API_ENDPOINT } = context.adapterConfig
+      return withApiKey(WS_API_ENDPOINT, API_CLIENT_KEY, API_CLIENT_SECRET)
+    },
+    handlers: {
+      message: (message) => {
+        if (Object.keys(message).length === 0) {
+          logger.debug('WS message is empty, skipping')
+          return []
+        }
+        return message.map((msg) => {
+          const base = baseFromIncludes[msg?.s] ?? msg?.s
+          wsTransport.setReverseMapping(msg.i, { base: msg.s, quote: 'USD' })
+          const result = {
+            params: { base, quote: 'USD' },
+            response: {
+              result: msg.price,
+              data: {
+                result: msg.price,
+              },
+              timestamps: {
+                providerIndicatedTimeUnixMs: new Date(msg.dt).getTime(),
+              },
+            },
+          }
+          return result
+        })
+      },
+    },
+    builders: {
+      subscribeMessage: (params) => {
+        const from = wsTransport.getReverseMapping(`${params.quote}${params.base}`)
```

Comment:
Unsure about the order here unless I'm missing something about the DP that requires this. I think we should be setting the reverse map here. And we should use it for getting the symbol to send in the subscription message

Reply:
the problem is that the message returns this value:
```
message [
        {
          s: 'USDCAD:CUR',
          i: 'USDCAD',
          pch: 0.26,
          nch: 0.00328,
          bid: 1.28778,
          ask: 1.28778,
          price: 1.28778,
          dt: 1659472542655,
          state: 'open',
          type: 'currency',
          dhigh: 1.2887,
          dlow: 1.2831,
          o: 1.28707,
          prev: 1.2845,
          topic: 'USDCAD'
        }
      ]
  ```    

 so I have no way of knowing what the base and quote are, apart from the fact that the base has a particular format 
 s : `quotebase:currency`
 
 so for this reason I did the reverse mapping `set` in the message and the `get` in the `subscribeMessage`, because I need that `s` particular format to subscribe
 
here is the example they give for ws: 
 https://github.com/tradingeconomics/tradingeconomics/blob/master/nodejs/Examples/stream-nodejs/app.js

Reply:
From what I'm seeing in the feed configurations for this adapter, I think we override the request to be in the `USDCAD:CUR` format. That's probably why the v2 version uses just the base for requests. So you could use the base as the key in `subscribeMessage` and then use `s` in the provider message to retrieve the params. Here's an example request.

```
{
    "from": "ETH",
    "overrides": {
      "tradingeconomics": {
        "ETH": "ETHUSD:CUR"
      }
    },
    "to": "USD"
}
```

My other concern is that we have to subscribe to the pair to retrieve the message from the provider in the first place. Or does Tradingeconomics start sending data for all symbols on connection? I know some DP like ncfx does that.

Reply:
If we're already overriding the request to only care about the base, the same should apply for the messages that come into the EA. There should be no need to use a reverse mapping. We can discuss this more in detail tomorrow